### PR TITLE
idempotent _clean_up() functions

### PR DESCRIPTION
**Pull request by graebm**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/task_scheduler.c [hunk 1]
```diff
@@ -57,24 +57,37 @@ static void s_run_all(struct aws_task_scheduler *scheduler, uint64_t current_tim
...
+        /* Execute all remaining tasks as CANCELED.
+         * Do this in a loop so that tasks scheduled by other tasks are executed */
+        while (aws_task_scheduler_has_tasks(scheduler, NULL)) {
+            s_run_all(scheduler, UINT64_MAX, AWS_TASK_STATUS_CANCELED);
+        }
     }
 
     aws_priority_queue_clean_up(&scheduler->timed_queue);
-    AWS_ZERO_STRUCT(scheduler);
+    AWS_ZERO_STRUCT(*scheduler);
```
Comment (feliperodri) [inline source/task_scheduler.c]: What about `timed_list` and `asap_list`?

File: source/task_scheduler.c [hunk 2]
```diff
@@ -57,24 +57,37 @@ static void s_run_all(struct aws_task_scheduler *scheduler, uint64_t current_tim
...
     scheduler->alloc = alloc;
     aws_linked_list_init(&scheduler->timed_list);
     aws_linked_list_init(&scheduler->asap_list);
-    return aws_priority_queue_init_dynamic(
-        &scheduler->timed_queue, alloc, DEFAULT_QUEUE_SIZE, sizeof(struct aws_task *), &s_compare_timestamps);
+    return AWS_OP_SUCCESS;
 }
 
 void aws_task_scheduler_clean_up(struct aws_task_scheduler *scheduler) {
     AWS_ASSERT(scheduler);
```
[the remark this answers is not in the document]
Comment (graebm) [reply source/task_scheduler.c]: No. That's the bug I'm fixing right now in aws_array_list_clean_up() 
The cleanup functions are supposed to be idempotent. It should be safe to call them multiple times). So the 2nd time you call clean_up() the datastructure *isn't* valid, and that should be ok.

[the remark this answers is not in the document]
Comment (graebm) [reply source/task_scheduler.c]: I just wrote a rant in chat about how I didn't want to write an aws_task_scheduler_is_wiped() function. I'll let the improvements be made when we proof this datastructure. I was just trying to quickly fix a bug I saw

File: source/task_scheduler.c [hunk 3]
```diff
@@ -57,24 +57,37 @@ static void s_run_all(struct aws_task_scheduler *scheduler, uint64_t current_tim
...
+            &scheduler->timed_queue, alloc, DEFAULT_QUEUE_SIZE, sizeof(struct aws_task *), &s_compare_timestamps)) {
+        return AWS_OP_ERR;
+    };
+
     scheduler->alloc = alloc;
     aws_linked_list_init(&scheduler->timed_list);
     aws_linked_list_init(&scheduler->asap_list);
-    return aws_priority_queue_init_dynamic(
-        &scheduler->timed_queue, alloc, DEFAULT_QUEUE_SIZE, sizeof(struct aws_task *), &s_compare_timestamps);
+    return AWS_OP_SUCCESS;
```
[the remark this answers is not in the document]
Comment (graebm) [reply source/task_scheduler.c]: did it

File: source/task_scheduler.c (hunk 1 above)
Comment (graebm) [reply source/task_scheduler.c]: The lists are empty at this point , and there is no aws_list_clean_up() function.

**Pull request merged**